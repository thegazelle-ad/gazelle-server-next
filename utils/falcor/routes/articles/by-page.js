import falcor from 'falcor';
import _ from 'lodash';

import { getPaginatedArticle } from './database-calls.sql';
import { database } from 'lib/db';

const $ref = falcor.Model.ref;

export const routes = [
  {
    /*
    Get articles by page (they are also in chronological order, articles['byPage'][pageLength][1][0]
    is the latest published article to the Ghost database). Only use positive integer page lengths
    and page numbers, and non-negative page indices. [{integers:indicesOnPage}] is logically
    redundant but needed for working properly with falcor. Normal falcorPath syntax would be:
    articles['byPage'][pageLength][pageIndex][{length: pageLength}]
    where {length: pageLength} makes use of falcor's range object.
    */
    route:
      "articles['byPage'][{integers:pageLengths}][{integers:pageIndices}][{integers:indicesOnPage}]",
    get: async pathSet => {
      validatePaginationQuery(
        pathSet.pageLengths,
        pathSet.pageIndices,
        pathSet.indicesOnPage,
      );
      // Because we validated that they want all indices on a page we no longer need that variable
      const pageLength = pathSet.pageLengths[0];
      const pageIndex = pathSet.pageIndices[0];

      const articles = await getPaginatedArticle(
        database,
        pageLength,
        pageIndex,
      );
      const results = articles.map(({ slug }, index) => ({
        path: ['articles', 'byPage', pageLength, pageIndex, index],
        value: $ref(['articles', 'bySlug', slug]),
      }));
      return results;
    },
  },
];

/**
 * @param {number} pageLengths
 * @param {number} pageIndices
 * @param {number} indicesOnPage
 * @throws if invalid, otherwise it'll finish silently
 */
function validatePaginationQuery(pageLengths, pageIndices, indicesOnPage) {
  if (pageLengths.length !== 1) {
    throw new Error(
      "We only allow passing a single pageLength at a time to articles['byPage']",
    );
  }
  if (pageLengths[0] < 1) {
    throw new Error('pageLength must be at least 1');
  }
  // It would be pretty simple to extend it to several pages, but as long as we don't use it
  // we can just as well make our SQL logic easier
  if (pageIndices.length !== 1) {
    throw new Error(
      "We only allow passing a single pageIndex at a time to articles['byPage']",
    );
  }
  if (pageIndices[0] < 0) {
    throw new Error('page indices must be non-negative');
  }
  if (indicesOnPage.length !== pageLengths[0]) {
    throw new Error(
      'You should always try fetching all the indices on a page, ' +
        'use {length: pageLength} for indices',
    );
  }
  if (_.uniq(indicesOnPage).length !== indicesOnPage.length) {
    throw new Error('All the indicesOnPage should be unique');
  }
  indicesOnPage.forEach(index => {
    if (index < 0 || index >= pageLengths[0]) {
      throw new Error('out of range indicesOnPage index passed');
    }
  });
}
