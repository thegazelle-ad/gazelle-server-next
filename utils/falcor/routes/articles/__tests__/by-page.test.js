import { routes } from '../by-page';

// database-calls is automatically mocked as seen in __mocks__
jest.enableAutomock();
jest.unmock('../by-page');
jest.unmock('lodash');
jest.unmock('falcor');

const mainRoutePath =
  "articles['byPage'][{integers:pageLengths}][{integers:pageIndices}][{integers:indicesOnPage}]";
const mainRoute = routes.find(x => x.route === mainRoutePath);

expect(mainRoute).not.toBeUndefined();

describe('articlesByPage falcor routes', () => {
  it('returns correct path values', async () => {
    const pageLength = 5;
    const pageIndex = 0;
    const pathValues = await mainRoute.get({
      pageLengths: [pageLength],
      pageIndices: [pageIndex],
      indicesOnPage: [0, 1, 2, 3, 4],
    });
    expect(pathValues).toMatchSnapshot();
  });

  it('throws on several pages requested', () =>
    expect(
      mainRoute.get({
        pageLengths: [1],
        pageIndices: [1, 2],
        indicesOnPage: [0],
      }),
    ).rejects.toThrow());

  it('throws on negative page index requested', () =>
    expect(
      mainRoute.get({
        pageLengths: [1],
        pageIndices: [-1],
        indicesOnPage: [0],
      }),
    ).rejects.toThrow());

  it('throws on negative pageLength requested', () =>
    expect(
      mainRoute.get({
        pageLengths: [-1],
        pageIndices: [2],
        indicesOnPage: [0],
      }),
    ).rejects.toThrow());

  it('throws on several page lengths requested', () =>
    expect(
      mainRoute.get({
        pageLengths: [1, 2],
        pageIndices: [2],
        indicesOnPage: [0],
      }),
    ).rejects.toThrow());

  it('throws when not all elements of a page were requested', () =>
    expect(
      mainRoute.get({
        pageLengths: [2],
        pageIndices: [2],
        indicesOnPage: [0],
      }),
    ).rejects.toThrow());

  it('throws when out of range element on page is requested', () =>
    expect(
      mainRoute.get({
        pageLengths: [2],
        pageIndices: [2],
        indicesOnPage: [0, 2],
      }),
    ).rejects.toThrow());

  it('throws when negative index on page is requested', () =>
    expect(
      mainRoute.get({
        pageLengths: [2],
        pageIndices: [2],
        indicesOnPage: [-1, 0],
      }),
    ).rejects.toThrow());

  it('throws if duplicate indices on page are requested', () =>
    expect(
      mainRoute.get({
        pageLengths: [2],
        pageIndices: [2],
        indicesOnPage: [0, 0],
      }),
    ).rejects.toThrow());
});
