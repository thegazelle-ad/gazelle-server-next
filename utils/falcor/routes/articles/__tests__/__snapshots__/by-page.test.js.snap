// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`articlesByPage falcor routes returns correct path values 1`] = `
Array [
  Object {
    "path": Array [
      "articles",
      "byPage",
      5,
      0,
      0,
    ],
    "value": Object {
      "$type": "ref",
      "value": Array [
        "articles",
        "bySlug",
        "slug-0",
      ],
    },
  },
  Object {
    "path": Array [
      "articles",
      "byPage",
      5,
      0,
      1,
    ],
    "value": Object {
      "$type": "ref",
      "value": Array [
        "articles",
        "bySlug",
        "slug-1",
      ],
    },
  },
  Object {
    "path": Array [
      "articles",
      "byPage",
      5,
      0,
      2,
    ],
    "value": Object {
      "$type": "ref",
      "value": Array [
        "articles",
        "bySlug",
        "slug-2",
      ],
    },
  },
  Object {
    "path": Array [
      "articles",
      "byPage",
      5,
      0,
      3,
    ],
    "value": Object {
      "$type": "ref",
      "value": Array [
        "articles",
        "bySlug",
        "slug-3",
      ],
    },
  },
  Object {
    "path": Array [
      "articles",
      "byPage",
      5,
      0,
      4,
    ],
    "value": Object {
      "$type": "ref",
      "value": Array [
        "articles",
        "bySlug",
        "slug-4",
      ],
    },
  },
]
`;
